### PR TITLE
[FLINK-15918][FLINK-15917] Uptime Metric not reset on Job Restart / Root Exception not shown in Web UI

### PR DESCRIPTION
**Pull request by GJL**
## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*

## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*

## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java [hunk 1]
```diff
@@ -210,6 +216,9 @@ private void restartTasksWithDelay(final FailureHandlingResult failureHandlingRe
 		final Set<ExecutionVertexVersion> executionVertexVersions =
 			new HashSet<>(executionVertexVersioner.recordVertexModifications(verticesToRestart).values());
 
+		transitionExecutionGraphState(JobStatus.RUNNING, JobStatus.RESTARTING);
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java]: extract method

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java [hunk 2]
```diff
@@ -223,6 +232,11 @@ private Runnable restartTasks(final Set<ExecutionVertexVersion> executionVertexV
 		return () -> {
 			final Set<ExecutionVertexID> verticesToRestart = executionVertexVersioner.getUnmodifiedExecutionVertices(executionVertexVersions);
 
+			verticesWaitingForRestart.removeAll(verticesToRestart);
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java]: extract method

**Comment by tillrohrmann** [comment]
I've started a separate build on travis-ci.com. You can track the progress here: https://travis-ci.com/flink-ci/flink/builds/147751241

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java [hunk 3]
```diff
@@ -596,6 +599,43 @@ public void suspendJobWillIncrementVertexVersions() {
...
+		final DefaultScheduler scheduler = createSchedulerAndStartScheduling(jobGraph);
+
+		final Iterator<ArchivedExecutionVertex> vertexIterator = scheduler.requestJob().getAllExecutionVertices().iterator();
+		final ExecutionAttemptID attemptId1 = vertexIterator.next().getCurrentExecutionAttempt().getAttemptId();
+		final ExecutionAttemptID attemptId2 = vertexIterator.next().getCurrentExecutionAttempt().getAttemptId();
+
+		scheduler.updateTaskExecutionState(new TaskExecutionState(jobGraph.getJobID(), attemptId1, ExecutionState.FAILED, new RuntimeException("expected")));
+		scheduler.updateTaskExecutionState(new TaskExecutionState(jobGraph.getJobID(), attemptId2, ExecutionState.FAILED, new RuntimeException("expected")));
+
+		taskRestartExecutor.triggerNonPeriodicScheduledTask();
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java]: We could check that before we process the reset execution task the `JobStatus` is already `RESTARTING`.

**Comment by GJL** [comment]
Test failure rate is acceptable.

**Comment by GJL** [comment]
```
2020-02-06 19:39:58,390 INFO  org.apache.flink.runtime.taskmanager.Task                     - Ensuring all FileSystem streams are closed for task Source: Custom Source (1/1) (e809b769c817e73fa009ec5055e883e2) [CANCELED]
2020-02-06 19:39:58,394 WARN  org.apache.flink.streaming.runtime.tasks.StreamTask           - Error while canceling task.
org.apache.flink.runtime.execution.CancelTaskException: Consumed partition PipelinedSubpartitionView(index: 0) of ResultPartition 20cd8853199f84e8a26339a17fe731e9@e809b769c817e73fa009ec5055e883e2 has been released.
	at org.apache.flink.runtime.io.network.partition.consumer.LocalInputChannel.getNextBuffer(LocalInputChannel.java:190)
	at org.apache.flink.runtime.io.network.partition.consumer.SingleInputGate.waitAndGetNextData(SingleInputGate.java:509)
	at org.apache.flink.runtime.io.network.partition.consumer.SingleInputGate.getNextBufferOrEvent(SingleInputGate.java:487)
	at org.apache.flink.runtime.io.network.partition.consumer.SingleInputGate.pollNext(SingleInputGate.java:475)
	at org.apache.flink.runtime.taskmanager.InputGateWithMetrics.pollNext(InputGateWithMetrics.java:75)
	at org.apache.flink.streaming.runtime.io.CheckpointedInputGate.pollNext(CheckpointedInputGate.java:125)
	at org.apache.flink.streaming.runtime.io.StreamTaskNetworkInput.emitNext(StreamTaskNetworkInput.java:133)
	at org.apache.flink.streaming.runtime.io.StreamOneInputProcessor.processInput(StreamOneInputProcessor.java:69)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.processInput(StreamTask.java:311)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMailboxLoop(MailboxProcessor.java:187)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.runMailboxLoop(StreamTask.java:487)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:470)
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:707)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:532)
	at java.lang.Thread.run(Thread.java:748)
2020-02-06 19:39:58,409 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor            - Un-registering task and sending final execution state CANCELED to JobManager for task Source: Custom Source (1/1) e809b769c817e73fa009ec5055e883e2.
2020-02-06 19:39:58,425 INFO  org.apache.flink.runtime.taskmanager.Task                     - Map -> Sink: Unnamed (1/1) (e2897fbac05b906ad242f5dd8e21fbbe) switched from CANCELING to CANCELED.
2020-02-06 19:39:58,425 INFO  org.apache.flink.runtime.taskmanager.Task                     - Freeing task resources for Map -> Sink: Unnamed (1/1) (e2897fbac05b906ad242f5dd8e21fbbe).
2020-02-06 19:39:58,426 INFO  org.apache.flink.runtime.taskmanager.Task                     - Ensuring all FileSystem streams are closed for task Map -> Sink: Unnamed (1/1) (e2897fbac05b906ad242f5dd8e21fbbe) [CANCELED]
2020-02-06 19:39:58,427 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor            - Un-registering task and sending final execution state CANCELED to JobManager for task Map -> Sink: Unnamed (1/1) e2897fbac05b906ad242f5dd8e21fbbe.
2020-02-06 19:39:58,451 INFO  org.apache.flink.runtime.taskexecutor.slot.TaskSlotTableImpl  - Free slot TaskSlot(index:0, state:ACTIVE, resource profile: ResourceProfile{cpuCores=1.0000000000000000, taskHeapMemory=384.000mb (402653174 bytes), taskOffHeapMemory=0 bytes, managedMemory=512.000mb (536870920 bytes), networkMemory=128.000mb (134217730 bytes)}, allocationId: 0fd30c7ec2d4e14753fe9e4803721b73, jobId: b3cc3c987e7aa8f99f5814b05d0057fe).
2020-02-06 19:39:58,452 INFO  org.apache.flink.runtime.taskexecutor.JobLeaderService        - Remove job b3cc3c987e7aa8f99f5814b05d0057fe from job leader monitoring.
2020-02-06 19:39:58,452 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor            - Close JobManager connection for job b3cc3c987e7aa8f99f5814b05d0057fe.
2020-02-06 19:39:58,455 INFO  org.apache.flink.runtime.taskexecutor.TaskExecutor            - Close JobManager connection for job b3cc3c987e7aa8f99f5814b05d0057fe.
2020-02-06 19:39:58,455 INFO  org.apache.flink.runtime.taskexecutor.JobLeaderService        - Cannot reconnect to job b3cc3c987e7aa8f99f5814b05d0057fe because it is not registered.
Checking for non-empty .out files...
No non-empty .out files.

[FAIL] 'State Migration end-to-end test from 1.6' failed after 0 minutes and 17 seconds! Test exited with exit code 0 but the logs contained errors, exceptions or non-empty .out files
```

Shouldn't be related?

**Comment by zhuzhurk** [comment]
> https://api.travis-ci.org/v3/job/646941244/log.txt

Looks this is not an issue. Just that the upstream task was canceled earlier and triggered a CancelTaskException of the downstream task which is still processing data on the same TM.

**Comment by GJL** [comment]
I was just about to post that this is not a problem. See https://issues.apache.org/jira/browse/FLINK-15751?focusedCommentId=17024366&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17024366

**Comment by tillrohrmann** [comment]
https://travis-ci.com/flink-ci/flink/builds/147754663 seems to pass as well.

Could we check whether we have a test ensuring that when we are in `RESTARTING` and calling `cancel` on the `ExecutionGraph` that we stop all tasks and eventually go into the state `CANCELED`?

**Comment by GJL** [comment]
@tillrohrmann  https://github.com/apache/flink/pull/11032/files#diff-71e4371a8ee17b6ce39a7022298ee228R627

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java [hunk 4]
```diff
@@ -596,6 +599,66 @@ public void suspendJobWillIncrementVertexVersions() {
...
+		final JobGraph jobGraph = singleJobVertexJobGraph(2);
+		final JobID jobid = jobGraph.getJobID();
+		final DefaultScheduler scheduler = createSchedulerAndStartScheduling(jobGraph);
+
+		final Iterator<ArchivedExecutionVertex> vertexIterator = scheduler.requestJob().getAllExecutionVertices().iterator();
+		final ExecutionAttemptID attemptId1 = vertexIterator.next().getCurrentExecutionAttempt().getAttemptId();
+		final ExecutionAttemptID attemptId2 = vertexIterator.next().getCurrentExecutionAttempt().getAttemptId();
+
+		scheduler.updateTaskExecutionState(new TaskExecutionState(jobid, attemptId1, ExecutionState.FAILED, new RuntimeException("expected")));
+		scheduler.cancel();
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java]: We should check that this call changes `attemptId2` to go into the `CANCELLING` state in order to make sure that the cancel call is being sent.

Comment (GJL) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java]: fixed

**Pull request closed without merging**